### ISSUE TITLE
ARM 32 ELF export fails for unchanged program: `Relocation at [...] exceeds file length`

### ISSUE THREAD

**Issue by Lanchon** [NationalSecurityAgency/ghidra#3696]
**Describe the bug**
ARM 32 ELF export fails for unchanged program with error `Relocation at [...] exceeds file length`

**To Reproduce**
1. Import attached program (ARM 32 ELF, is open source, part of DD-WRT).
2. Try to export the program: a window pops up with the error.

**Expected behavior**
The unchanged program should be successfully exported.

**Attachments**
[libutils.so.tar.gz](https://github.com/NationalSecurityAgency/ghidra/files/7640141/libutils.so.tar.gz)

**Environment (please complete the following information):**
 - OS: linux mint 19.3
 - Java Version: openjdk 11
 - Ghidra Version: 10.1-BETA
 - Ghidra Origin: official

**Comment by ryanmkurtz** [comment]
Thanks we are taking a look.

**Comment by ryanmkurtz** [comment]
There was a relocation that was only partially file-backed, so that case had to be handled.  The fix is working it's way through to you.

**Comment by Lanchon** [comment]
thanks so much for the immediate help!

no way to get my hands on a test build, is there?

**Comment by ryanmkurtz** [comment]
No, we don't distribute test builds.  If you are able to build Ghidra on your own I can tell you what you need to change.

**Comment by Lanchon** [comment]
thanks! i'm downloading the tree and deps as i write.

**Comment by Lanchon** [comment]
i was able to build Ghidra master :)

> I can tell you what you need to change

please!

**Comment by ryanmkurtz** [comment]
In this zip file is a new `AbstractLoaderExporter.java` file that you can drop in where the old one was at 
```
./Ghidra/Features/Base/src/main/java/ghidra/app/util/exporter/AbstractLoaderExporter.java
```

[AbstractLoaderExporter.zip](https://github.com/NationalSecurityAgency/ghidra/files/7645209/AbstractLoaderExporter.zip)

**Comment by Lanchon** [comment]
thanks sooo much!!

for documentation purposes...
```diff
diff --git a/Ghidra/Features/Base/src/main/java/ghidra/app/util/exporter/AbstractLoaderExporter.java b/Ghidra/Features/Base/src/main/java/ghidra/app/util/exporter/AbstractLoaderExporter.java
index e2e51654f..901e9e8e8 100644
--- a/Ghidra/Features/Base/src/main/java/ghidra/app/util/exporter/AbstractLoaderExporter.java
+++ b/Ghidra/Features/Base/src/main/java/ghidra/app/util/exporter/AbstractLoaderExporter.java
@@ -25,9 +25,11 @@ import ghidra.app.util.opinion.Loader;
 import ghidra.framework.model.DomainObject;
 import ghidra.program.database.mem.AddressSourceInfo;
 import ghidra.program.database.mem.FileBytes;
+import ghidra.program.model.address.Address;
 import ghidra.program.model.address.AddressSetView;
 import ghidra.program.model.listing.Program;
 import ghidra.program.model.mem.Memory;
+import ghidra.program.model.mem.MemoryBlockSourceInfo;
 import ghidra.program.model.reloc.Relocation;
 import ghidra.util.Conv;
 import ghidra.util.HelpLocation;
@@ -88,24 +90,32 @@ public abstract class AbstractLoaderExporter extends Exporter {
                }
                
                // Undo relocations in the temp file
-               // NOTE: not all relocations are file-backed
+               // NOTE: not all relocations are file-backed, and some are only partially file-backed
                String error = null;
                try (RandomAccessFile fout = new RandomAccessFile(tempFile, "rw")) {
                        Iterable<Relocation> relocs = () -> program.getRelocationTable().getRelocations();
                        for (Relocation reloc : relocs) {
-                               AddressSourceInfo info = memory.getAddressSourceInfo(reloc.getAddress());
-                               if (info == null) {
+                               Address addr = reloc.getAddress();
+                               AddressSourceInfo addrSourceInfo = memory.getAddressSourceInfo(addr);
+                               if (addrSourceInfo == null) {
                                        continue;
                                }
-                               long offset = info.getFileOffset();
-                               byte[] bytes = reloc.getBytes();
+                               long offset = addrSourceInfo.getFileOffset();
                                if (offset >= 0) {
-                                       if (offset + bytes.length > fout.length()) {
+                                       MemoryBlockSourceInfo memSourceInfo = addrSourceInfo.getMemoryBlockSourceInfo();
+                                       byte[] bytes = reloc.getBytes();
+                                       int len;
+                                       for (len = bytes.length; len > 1; len--) {
+                                               if (memSourceInfo.contains(addr.add(len - 1))) {
+                                                       break;
+                                               }
+                                       }
+                                       if (offset + len > fout.length()) {
                                                error = "Relocation at " + reloc.getAddress() + " exceeds file length";
                                                break;
                                        }
                                        fout.seek(offset);
-                                       fout.write(bytes);
+                                       fout.write(bytes, 0, len);
                                }
                        }
                }
```

**Comment by Lanchon** [comment]
@ryanmkurtz just letting you know that the fix worked and you are absolutely awesome for sharing the early fix. i was able to pinpoint a longstanding issue affecting some routers thanks to your help. you rock!!

**Comment by ryanmkurtz** [comment]
Great! I'm going to reopen this and let it get closed automatically by the real commit that will eventually make its way to GitHub.

**Comment by ryanmkurtz** [comment]
Was actually closed by 67569a918f020020ebb61faaff00599b1ab46709.  Note the final solution was a little different than the above code.
